Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift

```diff
@@ -129,38 +132,123 @@ public class DefaultAbly: AblyCommon {
+            self.detachFromChannel(channel: channel) { detachError in
+                guard let error = detachError else { return }
+                throw AblyError.connectionError(errorInfo: error)
+            }
+        }
+    }
+    
+    private func leavePresence(channel:AblySDKRealtimeChannel, presenceData: PresenceData?, completion : @escaping (ARTErrorInfo?) throws -> Void){
         let presenceDataJSON: Any?
         if let presenceData = presenceData {
             presenceDataJSON = self.presenceDataJSON(data: presenceData)
         } else {
             presenceDataJSON = nil
         }
         
-        channelToRemove.presence.leave(presenceDataJSON) { [weak self] error in
-            guard let error = error else {
-                self?.logHandler?.d(message: "\(String(describing: Self.self)): Left channel [id: \(trackableId)] presence successfully", error: nil)
-                channelToRemove.presence.unsubscribe()
-                channelToRemove.unsubscribe()
-                
-                channelToRemove.detach { [weak self] detachError in
-                    guard let error = detachError else {
-                        self?.channels.removeValue(forKey: trackableId)
-                        completion(.success(true))
-                        
-                        return
-                    }
-                    self?.logHandler?.e(message: "\(String(describing: Self.self)): Error during detach channel [id: \(trackableId)] presence", error: error)
-                    completion(.failure(error.toErrorInformation()))
-                }
-                
-                return
+         channel.presence.leave(presenceDataJSON){ errorInfo in
+             //This is to maintain closure signature
+             do{
+                 try completion(errorInfo)
+             }catch{}
+        }
+
+    }
+    
+    private func detachFromChannel(channel:AblySDKRealtimeChannel, completion : @escaping (ARTErrorInfo?) throws -> Void) {
+        channel.detach { errorInfo in
+            do{
+                try completion(errorInfo)
+            }catch{}
+        }
+    }
+
+    /**
+     * Performs the [operation] and if a "connection resume" exception is thrown it waits for the [channel] to
+     * reconnect and retries the [operation], otherwise it rethrows the exception. If the [operation] fails for
+     * the second time the exception is rethrown no matter if it was the "connection resume" exception or not.
+     */
+    private func performChannelOperationWithRetry(channel:AblySDKRealtimeChannel, operation : (AblySDKRealtimeChannel) throws ->Void) throws {
+      do {
+          logHandler?.w(message: "Trying to perform an operation on a suspended channel \(channel.name), waiting for the channel to be reconnected", error: nil)
+          try waitForChannelReconnection(channel: channel)
+          try operation(channel)
+      } catch  AblyError.connectionError(let errorInfo) {
```

Comment:
Why do you call `waitForChannelReconnection()` before doing the operation for the first time? :thinking: Also the log message seems odd here, it looks like some copy-paste error :wink:

Reply:
> @ikbalkaya I was wondering why you created #366? I think that this feature is sufficiently complicated that I would feel much more confident about merging it if we had tests.

I was about to leave for the pod work so I wanted to leave something reviewable. But now given I am working on this - I'll add tests for this PR

Reply:
2771dd37dd8400ff10e833e675f3394126ec4c22